C error from optional anonymous functions and `if str_var == 'some_string'` comparison

### Describe the bug

Bug is generated when setting an option field that is an anonymous function. The error only appears when using the if statement below. Doing `println(value == 'unlimited')` works.

### Reproduction Steps

```v
// CycleMenuItem is a menu item that can be cycled through by pressing a button.
//
pub struct CycleMenuItem {
mut:
	label  string
	values []string
	on     CycleMenuItemEvents
}

// CycleMenuItemEvents is a set of events that can be invoked by a
// `CycleMenuItem`.
[params]
pub struct CycleMenuItemEvents {
__global:
	click ?fn (string)
	cycle ?fn (string)
}

// CycleMenuItem.new creates a new `CycleMenuItem` with the given label, values,
// and function to invoke when the item is selected.
[inline]
pub fn CycleMenuItem.new(label string, values []string, events CycleMenuItemEvents) CycleMenuItem {
	return CycleMenuItem{
		label: label
		values: values
		on: events
	}
}

fn main() {
	_ := CycleMenuItem.new('FPS', ['30', '60', '90', '120', '144', '165', 'unlimited'],
		click: fn (value string) {
			if value == 'unlimited' {
			}
		}
	)
}

```

### Expected Behavior

Expected code to compile with no output.

### Current Behavior

```
C:/Users/imado/AppData/Local/Temp/v_0/test.4352526411636125798.tmp.c:500: warning: WINVER redefined
C:/Users/imado/AppData/Local/Temp/v_0/test.4352526411636125798.tmp.c:6816: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/imado/AppData/Local/Temp/v_0/test.4352526411636125798.tmp.c:12588: error: cannot convert 'typedef struct _option_anon_fn_string' to 'int'
builder error: 
==================
C error. This should never happen.
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V full version: V 0.4.1 68cbf27.24278d8

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.1 68cbf27.24278d8
OS: windows, Microsoft Windows 11 Pro v22621 64-bit
Processor: 16 cpus, 64bit, little endian, 

getwd: C:\frostbyte
vexe: C:\Users\imado\v\v.exe
vexe mtime: 2023-09-29 19:06:29

vroot: OK, value: C:\Users\imado\v
VMODULES: OK, value: C:\Users\imado\.vmodules
VTMP: OK, value: C:\Users\imado\AppData\Local\Temp\v_0

Git version: git version 2.33.1.windows.1
Git vroot status: weekly.2023.39-20-g24278d82 (1 commit(s) behind V master)
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,
operable program or batch file.

thirdparty/tcc status: thirdparty-windows-amd64 e90c2620
```
> [!NOTE]
> You can vote for this issue using the 👍 reaction. More votes increase the issue's priority for developers.
>
> Take into account that only the 👍 reaction counts as a vote.
> Only reactions to the issue itself will be counted as votes, not comments.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -1981,6 +1981,7 @@ fn (mut g Gen) expr_with_tmp_var(expr ast.Expr, expr_typ ast.Type, ret_typ ast.T
 		} else {
 			g.writeln(' }, (${c.result_name}*)(&${tmp_var}), sizeof(${styp}));')
 		}
+		g.set_current_pos_as_last_stmt_pos()
 	}
 
 	g.write(stmt_str)

--- a/vlib/v/tests/struct_init_with_multi_option_fn_type_test.v
+++ b/vlib/v/tests/struct_init_with_multi_option_fn_type_test.v
@@ -1,0 +1,43 @@
+// CycleMenuItem is a menu item that can be cycled through by pressing a button.
+pub struct CycleMenuItem {
+mut:
+	label  string
+	values []string
+	on     CycleMenuItemEvents
+}
+
+// CycleMenuItemEvents is a set of events that can be invoked by a
+// `CycleMenuItem`.
+[params]
+pub struct CycleMenuItemEvents {
+__global:
+	click ?fn (string) bool
+	cycle ?fn (string) bool
+}
+
+// CycleMenuItem.new creates a new `CycleMenuItem` with the given label, values,
+// and function to invoke when the item is selected.
+[inline]
+pub fn CycleMenuItem.new(label string, values []string, events CycleMenuItemEvents) CycleMenuItem {
+	return CycleMenuItem{
+		label: label
+		values: values
+		on: events
+	}
+}
+
+fn test_struct_init_with_multi_option_fn_type() {
+	item := CycleMenuItem.new('FPS', ['30', '60', '90', '120', '144', '165', 'unlimited'],
+		click: fn (value string) bool {
+			if value == 'unlimited' {
+				return true
+			} else {
+				return false
+			}
+		}
+	)
+	func := item.on.click?
+	ret := func('unlimited')
+	println(ret)
+	assert ret
+}

